### ISSUE TITLE
Slow category reload

### ISSUE THREAD

**Issue by lucaderi** [ntop/ntopng#2982]
Not sure why categories need to be reloaded though

22/Oct/2019 21:05:28 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.644,enp6s0f1.644]
22/Oct/2019 21:05:18 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.22,enp6s0f1.22]
22/Oct/2019 21:05:08 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.21,enp5s0f1.21]
22/Oct/2019 21:04:57 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.40,enp6s0f1.40]
22/Oct/2019 21:04:47 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.250,enp5s0f1.250]
22/Oct/2019 21:04:37 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.3686,enp6s0f1.3686]
22/Oct/2019 21:04:27 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.67,enp5s0f1.67]
22/Oct/2019 21:04:17 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.150,enp6s0f1.150]
22/Oct/2019 21:04:07 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.125,enp5s0f1.125]
22/Oct/2019 21:03:57 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.121,enp6s0f1.121]
22/Oct/2019 20:07:28 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.644,enp6s0f1.644]
22/Oct/2019 20:07:18 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.22,enp6s0f1.22]
22/Oct/2019 20:07:08 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.21,enp5s0f1.21]
22/Oct/2019 20:06:57 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.40,enp6s0f1.40]
22/Oct/2019 20:06:47 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.250,enp5s0f1.250]
22/Oct/2019 20:06:36 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.3686,enp6s0f1.3686]
22/Oct/2019 20:06:26 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.67,enp5s0f1.67]
22/Oct/2019 20:06:15 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.150,enp6s0f1.150]
22/Oct/2019 20:06:05 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.125,enp5s0f1.125]
22/Oct/2019 20:05:55 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.121,enp6s0f1.121]
22/Oct/2019 19:05:18 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.644,enp6s0f1.644]
22/Oct/2019 19:05:08 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.22,enp6s0f1.22]
22/Oct/2019 19:04:58 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.21,enp5s0f1.21]
22/Oct/2019 19:04:48 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.40,enp6s0f1.40]
22/Oct/2019 19:04:38 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.250,enp5s0f1.250]
22/Oct/2019 19:04:28 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.3686,enp6s0f1.3686]
22/Oct/2019 19:04:17 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.67,enp5s0f1.67]
22/Oct/2019 19:04:07 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.150,enp6s0f1.150]
22/Oct/2019 19:03:57 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp5s0f0.125,enp5s0f1.125]
22/Oct/2019 19:03:47 [LuaEngine.cpp:1868] ERROR: Interface didn't reload custom categories on time [iface: enp6s0f0.121,enp6s0f1.121]
2

**Comment by emanuele-f** [comment]
Categories are loaded/reloaded into each network interface when lists are refreshed as each interface holds a different nDPI struct. Did this happen during startup?

**Comment by lucaderi** [comment]
No during normal operations. If you look at logs they happen periodically, every hour.

**Comment by emanuele-f** [comment]
Please check out d1d789ea881117a64b72dfebdba630a609e8a7f3

**Comment by lucaderi** [comment]
Still bugged
![Screen Shot 2019-10-25 at 08 36 14](https://user-images.githubusercontent.com/4493366/67548661-a6af1c80-f702-11e9-9b3c-0e52a93d8d41.png)

**Comment by lucaderi** [comment]
What is the parameter I can set for maximum reload time?
Why ntop complains so much? If it takes lets say 1 min what's wrong it with?
Furthermore whenever there is nothing to do as in this case no reload is necessary

**Comment by emanuele-f** [comment]
Unnecessary reload has been fixed in 943012e8c3c8a123c69c6bb3d552dc6dcced836d . Regarding the reload time, the 10 seconds are used as a worst case deadline, so ntopng is not expected to take so much time to reload them. I've tried to reproduce it with different setups (pcap interface, pf_ring interface, dummy interface) however could not reproduce it. 

After commit a41278f4a39c259c7db2d9f6600566748bae8adf, please run ntopng with `-v5 | grep "custom categories"` and report the output after clicking "update now" for a list on the "Category Lists" page.

```
28/Oct/2019 12:32:35 [NetworkInterface.cpp:2488] Going to reload custom categories [iface: wlan0]
28/Oct/2019 12:32:36 [NetworkInterface.cpp:2490] Reload custom categories completed [iface: wlan0]
```

**Comment by emanuele-f** [comment]
The discussed rework has been implemented in 378a1e2cd8b71feee71721eacb87b31c6c0a135f . Now a separate ndpi_struct is populated with the categories and then swapped. The old structure is cleaned up after about 15 seconds. Successive reloads are delayed until at least 10 seconds have passed to make the shadow pointer work correctly.
